File: rmt/src/test/java/de/flower/test/AbstractIntegrationTests.scala
package de.flower.test

import org.springframework.test.context.transaction.TransactionalTestExecutionListener
import org.springframework.transaction.annotation.Transactional
import org.springframework.test.context.{TestExecutionListeners, ContextConfiguration}
import org.springframework.test.context.testng.AbstractTestNGSpringContextTests
import org.springframework.beans.factory.annotation.Autowired
import de.flower.rmt.service.IMyTeamManager
import org.testng.annotations.AfterMethod
import javax.persistence.{TransactionRequiredException, PersistenceContext, EntityManager}
import org.scalatest.Assertions
import org.springframework.transaction.PlatformTransactionManager
import de.flower.rmt.repository.IMyTeamRepo
import org.hibernate.AssertionFailure
import org.slf4j.{LoggerFactory, Logger}

/**
 *
 * @author oblume
 */
@TestExecutionListeners(Array(classOf[TransactionalTestExecutionListener]))
@Transactional
@ContextConfiguration(locations = Array("/applicationContext-test.xml"))
class AbstractIntegrationTests extends AbstractTestNGSpringContextTests with Assertions {

    protected var log: Logger = LoggerFactory.getLogger(this.getClass());


    /**************************************************************************/
    /*                        Daos                                            */
    /**************************************************************************/

    @Autowired
    protected var myTeamRepo: IMyTeamRepo = _


    /**************************************************************************/
    /*                        Services                                        */
    /**************************************************************************/

    @Autowired
    protected var myTeamManager: IMyTeamManager = _


    @PersistenceContext
    protected var em: EntityManager = _

    @Autowired
    protected var transactionManager: PlatformTransactionManager = _

    /**
     * Sync hibernate with database. This is done inside the transaction shortly before commiting/rollbacking the transaction.
     *
     * @param method the method
     */
    @AfterMethod
    def syncDatabase(method: java.lang.reflect.Method) {
        log info "Flushing EM"
        try {
            em.flush();
        } catch {
            case e: AssertionFailure => {
                log info "Test had previous hibernate errors. Cannot sync with database."
            }
            case e: TransactionRequiredException => {
                log info "Test " + method.getName + " is not transactional."
            }
            case e: RuntimeException => {
                var message: String = "Testmethod " + method.getName() +
                        " failed. Fix this testcase cause all following tests will be skipped after this error!";
                log.error(message, e);
                throw new RuntimeException(message, e);
            }
        }
    }


    /*
    @AfterMethod(alwaysRun = true)
    def syncDatabase(java.lang.reflect.Method method) {
        Session session;
        try {
            session = getHibernateSession();
        } catch (HibernateException e) {
            log.info("Unit test " + method.getName() + " is not transactional.");
            return;
        }
        if (session != null) {
            try {
                log.info("Flushing session!");
                session.flush();
            } catch (HibernateException e) {
                String message = "Testmethod " + method.getName()
                        + " failed. Fix this testcase cause all following tests will be skipped after this error!";
                log.error(message, e);
                throw new RuntimeException(message, e);
            }
        }
        disableHibernateDebugging();
    }
    */

}

File: rmt/src/test/java/de/flower/rmt/test/AbstractIntegrationTests.java
package de.flower.rmt.test;

/**
 * @author flowerrrr
 */

import ch.qos.logback.classic.spi.ILoggingEvent;
import de.flower.common.service.security.IPasswordGenerator;
import de.flower.common.test.mock.IListAppender;
import de.flower.common.test.mock.LogBackListAppender;
import de.flower.rmt.repository.IClubRepo;
import de.flower.rmt.repository.IEventRepo;
import de.flower.rmt.repository.ITeamRepo;
import de.flower.rmt.repository.IUserRepo;
import de.flower.rmt.service.*;
import de.flower.rmt.service.geocoding.IGeocodingService;
import de.flower.rmt.service.mail.IMailService;
import de.flower.rmt.service.mail.INotificationService;
import de.flower.rmt.service.mail.ITemplateService;
import de.flower.rmt.service.security.ISecurityService;
import org.apache.commons.lang3.Validate;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.security.authentication.TestingAuthenticationToken;
import org.springframework.security.core.Authentication;
import org.springframework.security.core.context.SecurityContextHolderStrategy;
import org.springframework.security.core.userdetails.UserDetails;
import org.springframework.security.core.userdetails.UserDetailsService;
import org.springframework.test.annotation.DirtiesContext;
import org.springframework.test.context.ContextConfiguration;
import org.springframework.test.context.testng.AbstractTestNGSpringContextTests;
import org.springframework.transaction.PlatformTransactionManager;
import org.testng.annotations.AfterMethod;
import org.testng.annotations.BeforeClass;
import org.testng.annotations.BeforeMethod;
import org.testng.annotations.Listeners;

import javax.persistence.EntityManager;
import javax.persistence.PersistenceContext;
import javax.sql.DataSource;
import javax.validation.Validator;
import java.lang.reflect.Method;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.Map;

/**
 * Base class for integration tests from app-layer downwards. no ui-support.
 * Use @link AbstractWicketIntegrationTests.
 *
 * @author flowerrrr
 */
@Listeners({de.flower.common.test.ExceptionLoggerTestListener.class})
@ContextConfiguration(locations = {"/applicationContext-test.xml"})
@DirtiesContext(classMode = DirtiesContext.ClassMode.AFTER_CLASS)
public class AbstractIntegrationTests extends AbstractTestNGSpringContextTests {

    protected Logger log = LoggerFactory.getLogger(this.getClass());

    /**************************************************************************/
    /*                        Daos                                            */

    /**
     * **********************************************************************
     */

    @Autowired
    protected ITeamRepo teamRepo;

    @Autowired
    protected IClubRepo clubRepo;

    @Autowired
    protected IUserRepo userRepo;

    @Autowired
    protected IEventRepo eventRepo;

    /**************************************************************************/
    /*                        Services                                        */

    /**
     * **********************************************************************
     */

    @Autowired
    protected ITeamManager teamManager;

    @Autowired
    protected IUserManager userManager;

    @Autowired
    protected IRoleManager roleManager;

    @Autowired
    protected IEventManager eventManager;

    @Autowired
    protected IPlayerManager playerManager;

    @Autowired
    protected IResponseManager responseManager;

    @Autowired
    protected IInvitationManager invitationManager;

    @Autowired
    protected UserDetailsService userDetailService;

    @Autowired
    protected IGeocodingService geocodingService;

    @Autowired
    protected IMailService mailService;

    @Autowired
    protected ITemplateService templateService;

    @Autowired
    protected INotificationService notificationService;

    @Autowired
    protected ISecurityService securityService;

    @Autowired
    protected SecurityContextHolderStrategy securityContextHolderStrategy;

    @Autowired
    protected Validator validator;

    @Autowired
    protected IPasswordGenerator passwordGenerator;

    @PersistenceContext
    protected EntityManager em;

    @Autowired
    protected PlatformTransactionManager transactionManager;

    @Autowired
    protected DataSource dataSource;

    protected IListAppender<ILoggingEvent> listAppender;

    @Autowired
    protected TestData testData;

    private Database db;

    public static final String testUserName = "manager-rmt@mailinator.com";

    @BeforeClass
    protected void initClass() {
        listAppender = LogBackListAppender.configureListAppender();
        // use db-unit to refresh the the test data
        Map properties = new HashMap();
        properties.put("http://www.dbunit.org/properties/datatypeFactory", new org.dbunit.ext.h2.H2DataTypeFactory());
        db = new Database(properties);
        db.setDataSource(dataSource);
    }

    @BeforeMethod
    protected void initTest() throws Exception {
        try {
            // reset database
            resetTestdata();
            resetListAppender();

            initializeSecurityContextWithTestUser();
        } catch (Exception e) {
            log.error(e.getMessage(), e);
            throw e;
        }
    }

    @AfterMethod
    protected void afterTest() throws Exception {
        try {
            // dump database to disk
            // full database export (cannot export users table: always a dbunit error wiht this table.)
            String[] tables = new String[]{"club", "team", "venue", "player", "event", "invitation"};
            db.export("src/test/database/data/export.xml", tables);
        } catch (Exception e) {
            log.error(e.getMessage(), e);
            throw e;
        }
    }

    /**
     * Initialize security context with test user.
     */
    protected void initializeSecurityContextWithTestUser() {
        // set the security context with a test user.
        Validate.notNull(userRepo.findByEmail(testUserName), "Make sure test user is present in test database.");
        UserDetails principal = userDetailService.loadUserByUsername(testUserName);
        Authentication authentication = new TestingAuthenticationToken(principal, principal.getPassword(), new ArrayList(principal.getAuthorities()));
        securityContextHolderStrategy.getContext().setAuthentication(authentication);
    }

    /**
     * Reset testdata. Most workflow unit tests are @NotTransactional annotated to have a more realistic transactional behaviour of the workflow-methods. That means we have to reset any test-data that
     * was modified during a testmethod.
     */
    private void resetTestdata() {
        log.info("Resetting test data in database.");

        db.deleteAll(Database.createDataSet("/data/test_data_delete.xml"));
        db.cleanInsert(Database.createDataSet("/data/test_data.xml"));

        testData.checkDataConsistency(em);
    }

    /**
     * Reset the internal buffer of the mock appender before each test method is started.
     */
    private void resetListAppender() {
        if (listAppender != null) {
            listAppender.reset();
        }
    }

    /**
     * Prints out "test started" message.
     * Requires junit 4.8 test runner that is able to process @Rule annotations.
     * Does not work with eclipse 3.5, which has junit 4.5 bundled.
     */
    @BeforeMethod(alwaysRun = true)
    public void testStarted(Method method) {
        log.info("***************************************************************");
        log.info("Test [" + method.getName() + "] started.");
        log.info("***************************************************************");
    }

    /**
     * Prints out "test finished" message.
     */
    @AfterMethod
    public void testFinished(Method method) {
        log.info("***************************************************************");
        log.info("Test [" + method.getName() + "()] finshed.");
        log.info("***************************************************************");
    }
}
